microsetta_admin/server.py

```diff
@@ -774,6 +775,191 @@ def scan():
+                        color_code = ['#' + ''.join(code)]
+                        while color_code[0] in dict:
+                            code = [random.choice(hex) for i in range(6)]
+                            color_code = ['#' + ''.join(code)]
+                        tmp = obj["values"][val].replace(' ', '-').lower()
+                        dict[tmp] = color_code[0]
+
+                if criteria == "Project" and obj["label"] == "Project":
+                    for val in obj["values"]:
+                        if not ('-' in obj["values"][val]):
+                            code = [random.choice(hex) for i in range(6)]
+                            color_code = ['#' + ''.join(code)]
+                            while color_code[0] in dict:
+                                code = [random.choice(hex) for i in range(6)]
+                                color_code = ['#' + ''.join(code)]
+                            dict[obj["values"][val]] = color_code[0]
+
+            if len(dict) > 0:
+                dict['None'] = "#000000"
+                session[criteria] = dict
+    else:
+        dict = session.get(criteria)
+    print(str(dict))
+    return dict
+
+
+def _compose_table(legends, type):
+    output_dict = {}
+    scanned_data = session.get('scan_data')
+    current_row = ""
+    current_cols = []
+
+    for rec in scanned_data:
+        status, data = APIRequest.get(
+            '/api/admin/search/samples/%s' % rec[5]
+        )
+
+        val = ""
+        proj = data["sample"]["sample_projects"][0]
+        status = data["sample"]["_latest_scan_status"]
+        src = data["sample"]["site"]
+
+        if type == "Type":
+            val = data["sample"]["site"]
+        elif type == "Status":
+            val = data["sample"]["_latest_scan_status"]
+        elif type == "Project":
+            val = data["sample"]["sample_projects"][0]
+
+        if current_row == "":
+            current_row = rec[4]
+            if val is None or val == "":
+                tmp = legends['None']
+                current_cols.append([rec[5], tmp, proj, status, src])
+            else:
+                current_cols.append([rec[5], legends[val], proj, status, src])
+        elif current_row == rec[4]:
+            if val is None:
+                tmp = legends['None']
+                current_cols.append([rec[5], tmp, proj, status, src])
+            else:
+                current_cols.append([rec[5], legends[val], proj, status, src])
+        else:
+            output_dict[current_row] = current_cols
+            current_cols = []
+            current_row = rec[4]
+            if val is None:
+                tmp = legends['None']
+                current_cols.append([rec[5], tmp, proj, status, src])
+            else:
+                current_cols.append([rec[5], legends[val], proj, status, src])
+
+    output_dict[current_row] = current_cols
+    return output_dict
+
+
+def post_bulk_scan():
```

Comment:
Can this function either be broken into three different functions for each "stage" or have the stages decomposed?

Reply:
divided for each stage and are now called from within bul_scan function as per the stage